In [2]:
import sys
sys.path.append('../src')


import nltk
nltk.download('punkt_tab')


from datasets import load_dataset
from collections import Counter
from data_loader import preprocess_text
from embeddings import load_pretrained_vocab


[nltk_data] Downloading package punkt_tab to
[nltk_data]     /Users/haoyangpang/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
/Users/haoyangpang/Desktop/Y3S1/SC4002 Natural Language Processing/assignment/venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
[nltk_data] Downloading package punkt to
[nltk_data]     /Users/haoyangpang/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [3]:
# Load the Rotten Tomatoes dataset
dataset = load_dataset("rotten_tomatoes")
train_texts = dataset['train']['text']
train_labels = dataset['train']['label']


In [4]:
# Preprocess all training texts to extract tokens
all_tokens = []
for text in train_texts:
    tokens = preprocess_text(text)
    all_tokens.extend(tokens)

# Build vocabulary from tokens
counter = Counter(all_tokens)
vocab = {'<PAD>': 0, '<UNK>': 1}
index = len(vocab)
for token in counter:
    vocab[token] = index
    index += 1

vocab_size = len(vocab)
print(f"Vocabulary size: {vocab_size}")


Vocabulary size: 15813


## Question 1(a): What is the size of the vocabulary formed from your training data?


Ans: the vocabulary size for the training data is 15813

In [5]:
# Load pre-trained GloVe vocabulary
embedding_file = '../glove.6B.300d.txt'  
pre_trained_vocab = load_pretrained_vocab(embedding_file)
print(f"Number of words in GloVe vocabulary: {len(pre_trained_vocab)}")


Number of words in GloVe vocabulary: 400000


In [6]:
# Identify OOV words in the training vocabulary
training_vocab_set = set(vocab.keys())
oov_words = training_vocab_set - pre_trained_vocab
num_oov = len(oov_words)
print(f"Number of OOV words in training data: {num_oov}")


Number of OOV words in training data: 550


## Question 1(b): How many OOV words exist in your training data?

Ans: there are 550 OOV words

## Question 1(c): Mitigating OOV Words

One effective strategy to handle OOV words is to initialize their embeddings with random vectors and update them during training. This allows the model to learn representations for OOV words based on the context in which they appear.

Check the updated version in embedding.py


In [7]:
# Save the vocabulary for later use
import pickle

with open('../vocab.pkl', 'wb') as f:
    pickle.dump(vocab, f)
